File: python_programs/sorting/bubbleSort.py
class Solution:
    def bubble_sort(self, nums):
        def pass_swap(sorted_index):  # sorted_index -> no. of indices sorted at the end of the array
            swaps = 0
            for i in range(len(nums) - sorted_index - 1):  
                if nums[i] > nums[i + 1]:
                    nums[i], nums[i + 1] = nums[i + 1], nums[i]
                    swaps += 1
            return swaps  # returns the no. of swaps

        for i in range(0, len(nums)):
            if not pass_swap(i):
                return                      # already sorted, return early


        


sol = Solution()
nums = list(map(int, input("Enter Array: ").split()))
sol.bubble_sort(nums)
print(nums)

File: c_programs/bubble_sort.c
#include<stdio.h>

int pass_and_swap(int arr[], int sorted_index, int size);

int main()
{
    int nums[100];
    int size;
    printf("Enter size: ");
    scanf("%d", &size);
    for(int i = 0; i < size; i++)
    {
        scanf("%d", &nums[i]);
    }

    for(int i = 0; i < size; i++)
    {
        int swaps = pass_and_swap(nums, i, size);
    }
    for(int i = 0; i < size; i++)
    {
        printf("%d ", nums[i]);
    }
}

int pass_and_swap(int arr[],int sorted_index, int size)
{
    int swaps = 0;
    for(int i = 0; i < size - sorted_index - 1; i++)
    {
        if (arr[i] > arr[i+1])
        {
            arr[i] = arr[i] + arr[i+1]; //swaps arr[i] and arr[i+1]
            arr[i+1] = arr[i] - arr[i+1];
            arr[i] = arr[i] - arr[i+1];
            swaps++;
        }
    }
    return swaps;
}
